### ISSUE TITLE
[Question]: how does llamaindex support large datasets?

### ISSUE THREAD

**Issue by lambda7xx** [run-llama/llama_index#13987]
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

I use the datasets from https://hotpotqa.github.io/ which is 500MB. I want to implement the iterative application. But when I load the datasets and test the query. After 10 minutes, the code is still running. The below is our code
```
from llama_index.core import VectorStoreIndex, get_response_synthesizer
from llama_index.core.retrievers import VectorIndexRetriever
from llama_index.core.query_engine import RetrieverQueryEngine
from llama_index.core.postprocessor import SimilarityPostprocessor

#export CUDA_VISIBLE_DEVICES=2
import torch
# from transformers import BitsAndBytesConfig

from llama_index.llms.huggingface import HuggingFaceLLM
from llama_index.core.prompts import PromptTemplate

from llama_index.core import Settings

from llama_index.core import VectorStoreIndex
from llama_index.core import  SimpleDirectoryReader
from llama_index.core.response.notebook_utils import display_response
from llama_index.core.indices.query.query_transform.base import (
    StepDecomposeQueryTransform,
)
import llama_index.core 
import time 

import time
# from llama_index.llms import HuggingFaceLLM

# llama_index.core.set_global_handler("simple")

llm = HuggingFaceLLM(
    model_name="microsoft/Phi-3-medium-128k-instruct",
    tokenizer_name="microsoft/Phi-3-medium-128k-instruct",
    query_wrapper_prompt=PromptTemplate("<s>[INST] {query_str} [/INST] </s>\n"),
    context_window=3900,
    max_new_tokens=256,
    model_kwargs={"torch_dtype": torch.bfloat16},
    generate_kwargs={"temperature": 0.2, "top_k": 5, "top_p": 0.95},
    device_map="auto",
    # trust_remote_code=True,
)

# from vllm import LLM, SamplingParams

# llm = LLM(
#     model="microsoft/Phi-3-medium-128k-instruct",
#     tensor_parallel_size=1,
#     max_model_len=5000,
#     trust_remote_code=True,
#     enforce_eager=True,
#     # GLM-4-9B-Chat-1M 如果遇见 OOM 现象，建议开启下述参数
#     # enable_chunked_prefill=True,
#     # max_num_batched_tokens=8192
# )


Settings.llm = llm

embeddings =  "local:BAAI/bge-small-en-v1.5" #"local:BAAI/bge-small-en-v1.5"

Settings.embed_model = embeddings

from load_data import read_json

qa_pairs = read_json() 

query0 = qa_pairs[0][0]
#load document 
documents = SimpleDirectoryReader("./data").load_data()
index = VectorStoreIndex.from_documents(documents=documents)
# build index
# index = VectorStoreIndex.from_documents(documents)

# configure retriever
retriever = VectorIndexRetriever(
    index=index,
    similarity_top_k=10,
)

# configure response synthesizer
response_synthesizer = get_response_synthesizer()

# assemble query engine
query_engine = RetrieverQueryEngine(
    retriever=retriever,
    response_synthesizer=response_synthesizer,
    node_postprocessors=[SimilarityPostprocessor(similarity_cutoff=0.7)],
)

# query
response = query_engine.query(query0)
print(response)

```

**Comment by lambda7xx** [comment]
```
#Multi-Step Query Engine https://docs.llamaindex.ai/en/stable/examples/query_transformations/SimpleIndexDemo-multistep.html
from llama_index.core import (
    VectorStoreIndex,
    SimpleDirectoryReader
)
from llama_index.llms.ollama import Ollama
from llama_index.core.embeddings import resolve_embed_model
from llama_index.core import SimpleDirectoryReader, Settings
from llama_index.postprocessor.cohere_rerank import CohereRerank
from llama_index.core import PromptTemplate
from llama_index.core import VectorStoreIndex, SimpleDirectoryReader
from IPython.display import Markdown, display

from llama_index.llms.huggingface import HuggingFaceLLM


# bge embedding model
Settings.embed_model = resolve_embed_model("local:BAAI/bge-small-en-v1.5")

llm = HuggingFaceLLM(
    model_name="THUDM/glm-4-9b-chat",
    tokenizer_name="THUDM/glm-4-9b-chat",
    query_wrapper_prompt=PromptTemplate("<s>[INST] {query_str} [/INST] </s>\n"),
    context_window=3900,
    max_new_tokens=256,
    # model_kwargs={"torch_dtype": torch.bfloat16},
    generate_kwargs={"temperature": 0.2, "top_k": 5, "top_p": 0.95},
    device_map="auto",
    # trust_remote_code=True,
)


# ollama
Settings.llm =llm


documents = SimpleDirectoryReader("./data").load_data()

index = VectorStoreIndex.from_documents(documents)


from llama_index.core.indices.query.query_transform.base import (
    StepDecomposeQueryTransform,
)


step_decompose_transform = StepDecomposeQueryTransform(llm=Settings.llm, verbose=True)

# index_summary = "Used to answer questions about the author"

from llama_index.core.query_engine import MultiStepQueryEngine

from load_data import read_json

qa_pairs = read_json() 

query0 = qa_pairs[0][0]

print(f"query0: {query0}")

query_engine = index.as_query_engine(llm=Settings.llm)
query_engine = MultiStepQueryEngine(
    query_engine=query_engine,
    query_transform=step_decompose_transform,
)
response_gpt4 = query_engine.query(
    query0
)

print(f"response_gpt4.response: {response_gpt4.response}")
# from IPython.display import HTML
# display(HTML(f'{response_gpt4.response}'))
```
I also run this code. it's still need 10 miniutes.

**Comment by logan-markewich** [comment]
hotpotqa is a huge dataset, and you are using local models

If you want to speed it up, you'll need either better hardware, or use APIs that are faster (i.e. text embedding inference and text generation inference would be one example)

**Comment by lambda7xx** [comment]
> hotpotqa is a huge dataset, and you are using local models
> 
> If you want to speed it up, you'll need either better hardware, or use APIs that are faster (i.e. text embedding inference and text generation inference would be one example)

My GPU is H100 GPU.

what do you mean that use ` text embedding inference and text generation inference`? I think I use the huggingface  text generation inference.

**Comment by lambda7xx** [comment]
does I need to use a vector db for my datasets? If yes, which one you recommend? thanks

**Comment by lambda7xx** [comment]
I also try the vllm, it's very slow.

**Comment by logan-markewich** [comment]
`text embedding inference and text generation inference`  -- yea I meant these, they are a more optimized runtime
- https://github.com/huggingface/text-embeddings-inference
- https://github.com/huggingface/text-generation-inference

You can also increase the embedding model batch size if the creation is a bottleneck
`HuggingFaceEmbedding(..., embed_batch_size=128)`

If you are using a vllm server (or any server based API), I recommend you can take advantage of async and concurrency as well. 

```
from llama_index.core.async_utils import run_jobs

jobs = [query_engine.aquery(query) for query in queries]

# here, workers is how many outgoing calls can be awaited at any given time, helps reduce load on the server
results = await run_jobs(jobs, workers=5, show_progress=True)
```

Note that if you are using models that run in process (like huggingface embeddings) and not over a server (i.e. text embedding inference), the call will be blocking, making async less useful

**Comment by lambda7xx** [comment]
The below are my new code. It's still very slow.

```

from llama_index.core import VectorStoreIndex, get_response_synthesizer
from llama_index.core.retrievers import VectorIndexRetriever
from llama_index.core.query_engine import RetrieverQueryEngine
from llama_index.core.postprocessor import SimilarityPostprocessor

#export CUDA_VISIBLE_DEVICES=2
import torch
# from transformers import BitsAndBytesConfig

from llama_index.llms.huggingface import HuggingFaceLLM
from llama_index.core.prompts import PromptTemplate

from llama_index.core import Settings

from llama_index.core import VectorStoreIndex
from llama_index.core import  SimpleDirectoryReader
from llama_index.core.response.notebook_utils import display_response
from llama_index.core.indices.query.query_transform.base import (
    StepDecomposeQueryTransform,
)
import llama_index.core 
import time 

import time


from llama_index.llms.vllm import Vllm

llm = Vllm(
    model="meta-llama/Llama-2-7b-hf",
    tensor_parallel_size=1,
    max_new_tokens=3000,
    vllm_kwargs={"swap_space": 1, "gpu_memory_utilization": 0.5},
    # messages_to_prompt=messages_to_prompt,
    # completion_to_prompt=completion_to_prompt,
)



Settings.llm = llm

embeddings =  "local:BAAI/bge-small-en-v1.5" #"local:BAAI/bge-small-en-v1.5"

from llama_index.embeddings.text_embeddings_inference import (
    TextEmbeddingsInference,
)

Settings.embed_model = TextEmbeddingsInference(
    model_name  = embeddings,
    embed_batch_size=128
)

from load_data import read_json

qa_pairs = read_json() 

query0 = qa_pairs[0][0]
#load document 
documents = SimpleDirectoryReader("./data").load_data()
index = VectorStoreIndex.from_documents(documents=documents)
# build index
# index = VectorStoreIndex.from_documents(documents)

# configure retriever
retriever = VectorIndexRetriever(
    index=index,
    similarity_top_k=10,
)

# configure response synthesizer
response_synthesizer = get_response_synthesizer()

# assemble query engine
query_engine = RetrieverQueryEngine(
    retriever=retriever,
    response_synthesizer=response_synthesizer,
    node_postprocessors=[SimilarityPostprocessor(similarity_cutoff=0.7)],
)

# query
response = query_engine.query(query0)
print(response)
```

**Comment by lambda7xx** [comment]
maybe I should try this https://docs.llamaindex.ai/en/stable/examples/vector_stores/FaissIndexDemo/

**Comment by lambda7xx** [comment]
> maybe I should try this https://docs.llamaindex.ai/en/stable/examples/vector_stores/FaissIndexDemo/

it still can't work even I use faiss.

```

from llama_index.core import VectorStoreIndex, get_response_synthesizer
from llama_index.core.retrievers import VectorIndexRetriever
from llama_index.core.query_engine import RetrieverQueryEngine
from llama_index.core.postprocessor import SimilarityPostprocessor
from llama_index.core import (
    SimpleDirectoryReader,
    load_index_from_storage,
    VectorStoreIndex,
    StorageContext,
)
#export CUDA_VISIBLE_DEVICES=2
import torch
# from transformers import BitsAndBytesConfig

from llama_index.llms.huggingface import HuggingFaceLLM
from llama_index.core.prompts import PromptTemplate

from llama_index.core import Settings

from llama_index.core import VectorStoreIndex
from llama_index.core import  SimpleDirectoryReader
from llama_index.core.response.notebook_utils import display_response
from llama_index.core.indices.query.query_transform.base import (
    StepDecomposeQueryTransform,
)
import llama_index.core 
import time 

import time

from llama_index.vector_stores.faiss import FaissVectorStore

from llama_index.llms.vllm import Vllm

llm = Vllm(
    model="meta-llama/Llama-2-7b-hf",
    tensor_parallel_size=1,
    max_new_tokens=3000,
    vllm_kwargs={"swap_space": 1, "gpu_memory_utilization": 0.5},
    # messages_to_prompt=messages_to_prompt,
    # completion_to_prompt=completion_to_prompt,
)



Settings.llm = llm

embeddings =  "local:BAAI/bge-small-en-v1.5" #"local:BAAI/bge-small-en-v1.5"

from llama_index.embeddings.text_embeddings_inference import (
    TextEmbeddingsInference,
)

Settings.embed_model = TextEmbeddingsInference(
    model_name  = embeddings,
    embed_batch_size=128
)

import faiss

# dimensions of text-ada-embedding-002
d = 1536
faiss_index = faiss.IndexFlatL2(d)

from load_data import read_json

qa_pairs = read_json() 

query0 = qa_pairs[0][0]
#load document 
documents = SimpleDirectoryReader("./data").load_data()

vector_store = FaissVectorStore(faiss_index=faiss_index)

storage_context = StorageContext.from_defaults(vector_store=vector_store)
index = VectorStoreIndex.from_documents(
    documents, storage_context=storage_context
)

query_engine = index.as_query_engine()


# index = VectorStoreIndex.from_documents(documents=documents)
# build index
# index = VectorStoreIndex.from_documents(documents)

# configure retriever
# retriever = VectorIndexRetriever(
#     index=index,
#     similarity_top_k=10,
# )

# # configure response synthesizer
# response_synthesizer = get_response_synthesizer()

# # assemble query engine
# query_engine = RetrieverQueryEngine(
#     retriever=retriever,
#     response_synthesizer=response_synthesizer,
#     node_postprocessors=[SimilarityPostprocessor(similarity_cutoff=0.7)],
# )

# query
response = query_engine.query(query0)
print(response)
```

**Comment by logan-markewich** [comment]
Faiss isn't the issue, the issue is bottlenecks for running the llm across many queries

**Comment by logan-markewich** [comment]
Async will help if you use a server tbh
